src/lightning/pytorch/loops/fit_loop.py

```diff
@@ -244,12 +244,25 @@ def setup_data(self) -> None:
         if trainer.datamodule is not None:
             allow_zero_length |= trainer.datamodule.allow_zero_length_dataloader_with_multiple_devices
 
+        limits = []
+        for dl in combined_loader.flattened:
+            # determine number of batches
+            length = len(dl) if has_len_all_ranks(dl, trainer.strategy, allow_zero_length) else float("inf")
+            num_batches = _parse_num_batches(stage, length, trainer.limit_train_batches)
+            limits.append(num_batches)
+
+        combined_loader.limits = limits
+
+        trainer.training = True
```

Comment:
It is necessary because  `_select_data_fetcher()` want's to access the training stage, but it has not been set here at the time `setup_data()` is called.

Reply:
https://github.com/Lightning-AI/lightning/actions/runs/6083368123/job/16503109807?pr=18441

Reply:
I see. It would probably be simpler to pass the "step_name" (`training_step`, `validation_step`, ...) to `_select_data_fetcher`. It only uses the trainer to know this.

Reply:
Are you requesting the change? Or is this hypothetical?

Reply:
If possible, yes. It could also be done in a follow-up

Reply:
I'd like to do it in a follow-up, because this function requires the stage anyway even if the step_fx_name is passed. I'd like to stay focused on the main objective of the PR.